File: student-backend/src/main/java/com/student/studentCrud/service/GradeService.java
package com.student.studentCrud.service;

import com.student.studentCrud.dto.GradeDto;
import com.student.studentCrud.util.ResponseStructure;

import java.util.List;
import java.util.Map;

public interface GradeService {

    GradeDto saveGrade(long rollNumber, String subject, double score);

    GradeDto updateGrade(GradeDto gradeDto);

    GradeDto deleteGrade(long gradeId);

    List<GradeDto> findAllGrade();

    GradeDto updateScore(long gradeId, double score);

    ResponseStructure<Map<String, Double>> findGradeByStudent(long rollNumber);

}

File: student-frontend/src/app/core/services/grade.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root',
})
export class GradeService {
  private baseUrl = 'http://localhost:8090/grade';

  constructor(private http: HttpClient) {}

  getGradeByStudent(rollNumber: number) {
    return this.http.get(
      `${this.baseUrl}/student/${rollNumber}`
    );
  }
  saveGrade(rollNumber: number, subject: string, score: number) {
  return this.http.post(
    `${this.baseUrl}/save`,
    null,
    { params: { rollNumber, subject, score } }
  );
}

deleteGrade(gradeId: number) {
  return this.http.delete(`${this.baseUrl}/delete/${gradeId}`);
}

updateScore(gradeId: number, score: number) {
  return this.http.patch(
    `${this.baseUrl}/update/score`,
    null,
    {
      params: {
        gradeId: gradeId,
        score: score
      }
    }
  );
}
}
